tests/nightly/test_tutorial.py

```diff
@@ -25,87 +25,116 @@
+        output_nb = open(output_file, mode='r')
         for line in output_nb:
             if "Warning:" in line:
-                fail_dict[tutorial_name] = "%s has warning." % (tutorial_name)
-                return
+                success = False
+                if tutorial_name in fail_dict:
+                    fail_dict[tutorial_name] += "\n"+line
+                else:
+                    fail_dict[tutorial_name] = "Warning:\n"+line
+        sys.stdout.write(' Elapsed time: {0:.2f}s '.format(time.time()-tick  ))
+        sys.stdout.write(' [{}] \n'.format('Success' if success else 'Failed'))
+        sys.stdout.flush()
 
 
 if __name__ == "__main__":
-    tutorial_dir = '../../docs/_build/html/tutorials/'
-    with open('test_tutorial_config.txt') as config_file:
-        tutorial_list = []
-        for line in config_file:
-            tutorial_list.append(line.lstrip().rstrip())
-            file_dir = tutorial_dir + line.lstrip().rstrip()
-            test_tutorial_nb(file_dir)
+    tutorial_dir = os.path.join('..','..','docs', '_build', 'html', 'tutorials')
+    tick = time.time()
+    
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--tutorial", help="tutorial to test, if not set, read from test_tutorial_config.txt")
+    parser.add_argument("--kernel", help="name of the jupyter kernel to use for the test")
+    parser.add_argument("--no-cache", help="clean the temp directory", action="store_true", dest="no_cache")
+    args = parser.parse_args()
+    
+
+    tutorial_list = []
+    if args.tutorial:
+        tutorial_list.append(args.tutorial)
+    else:
+        with open('test_tutorial_config.txt') as config_file:
+            for line in config_file:
+                tutorial_list.append(line.lstrip().rstrip())
+    
+    temp_dir = 'tmp_notebook'
+    if args.no_cache:
+        print("Cleaning and setting up temp directory '{}'".format(temp_dir))
+        shutil.rmtree(temp_dir, ignore_errors=True)
+    
+    kernel = args.kernel if args.kernel else None
+    
+    for tutorial in tutorial_list:
+        file_dir = os.path.join(*([tutorial_dir]+tutorial.split('/')))
+        working_dir = os.path.join(*([temp_dir]+tutorial.split('/')))
+        if not os.path.isdir(working_dir):
+            os.makedirs(working_dir)
+        test_tutorial_nb(file_dir, working_dir, kernel)
 
-        fail_num = len(fail_dict)
-        success_num = len(tutorial_list) - fail_num
-        print "Test Summary Start"
-        print "%d tutorials tested:" % (len(tutorial_list))
-        for tutorial in tutorial_list:
-            print tutorial
-        print "\n%d tests failed:" % (fail_num)
-        for tutorial, msg in fail_dict.items():
-            print tutorial + ":"
-            print msg
-        print "Test Summary End"
-        print "Stats start"
-        print "[Passed: %d of %d]" % (success_num, len(tutorial_list))
-        print "Stats end"
+    fail_num = len(fail_dict)
```

Comment:
fail_dict is never written

Reply:
it is declared at the top of the file, and written to here: 

```
+                if tutorial_name in fail_dict:
+                    fail_dict[tutorial_name] += "\n"+line
+                else:
+                    fail_dict[tutorial_name] = "Warning:\n"+line
```